### PR TITLE
Fix Issue 17181 - Local imports in templates should be added to impor…

### PR DESCRIPTION
**Pull request by RazvanN7**
…ted modules list of module that instantiated it

I don't know how to test this. @andralex maybe check your PR against this patch?

### CODE REVIEW

**Review by thewilsonator — COMMENTED** [review]
Tests?

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -1380,8 +1380,19 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (imp.mod)
         {
             // Modules need a list of each imported module
-            //printf("%s imports %s\n", sc._module.toChars(), imp.mod.toChars());
-            sc._module.aimports.push(imp.mod);
+
+            // if inside a template instantiation, the instantianting
+            // module gets the import.
```
Comment (WalterBright) [inline src/dmd/dsymbolsem.d]:
```
auto m = (sc.minst && sc.tinst) ? sc.minst : sc._module;
m.aimports.push(imp.mod);
```
4 lines of code to 2.

Comment (WalterBright) [reply src/dmd/dsymbolsem.d]: Might want to add a link to 17181 for rationale.

**Comment by RazvanN7** [comment]
@WalterBright done.
@thewilsonator Added test.

**Comment by RazvanN7** [comment]
>I suspect there are more complications: if a template is instantiated by mutliple modules, the import is added to the module that happens to be analyzed first. Other modules that reuse the same template instance need to have a dependency on the import, too.

I added a fix, but I have no idea how to test this case. I don't even think it's possible.

**Comment by wilzbach** [comment]
> > I suspect there are more complications: if a template is instantiated by mutliple modules, the import is added to the module that happens to be analyzed first. Other modules that reuse the same template instance need to have a dependency on the import, too.
> 
> I added a fix, but I have no idea how to test this case. I don't even think it's possible.

Custom shellscript with multiple EXTRA_FILES?
In the worst case you could always grep on the created (and dumped) object file.
There are a few shell tests which already do similar things.

**Comment by RazvanN7** [comment]
>Custom shellscript with multiple EXTRA_FILES?
In the worst case you could always grep on the created (and dumped) object file.
There are a few shell tests which already do similar things.

You can't see imports in object files. We are talking aboud seeing an internal list in the compiler.

**Comment by jacob-carlborg** [comment]
> We are talking aboud seeing an internal list in the compiler.

Use the DMD frontend as a library. But you need to install Dub for that.

**Comment by wilzbach** [comment]
> You can't see imports in object files. We are talking aboud seeing an internal list in the compiler.

Maybe grep on `dmd -v` then?

> But you need to install Dub for that.

Though the existing dub tests for the DMD frontend as a library are already tested on CircleCi.

**Comment by jacob-carlborg** [comment]
> Though the existing dub tests for the DMD frontend as a library are already tested on CircleCi.

But not on all platforms.

**Comment by RazvanN7** [comment]
I think that we can merge this as is. We have a test for the original problem and its observable effect. The problem that @rainers specified, even though it is solved by this PR it didn't have any observable effects.

**Review by andralex — APPROVED** [review]
LGTM. @WalterBright? @wilzbach? Any others?

**Review by rainers — APPROVED** [review]
LGTM. A test case that shows the dependency issue with reusing the template instance should be feasible with seperate compilation (especially through  library) but probably not worth the trouble.

An interesting change is that a module that only contains templates with imports, but no global imports, now no longer needs to be compiled to an object file as it should never appear in a dependency chain itself. Maybe worth a test case...

**Comment by rainers** [comment]
> You can't see imports in object files.

You can if you dump the data segments, too. You'll find the imports that are in the dependency chain in the `__ModuleInfo` section. See the ASM of https://run.dlang.io/is/6rUcxc

**Comment by ZombineDev** [comment]
@RazvanN7 are you up to adding a test, as per @rainers's comment?

**Pull request merged**